### DISCUSSION TITLE
OptionsValidationException: The 'Instance' option must be provided

### DISCUSSION THREAD

**Discussion by eloekset** [dotnet/aspnetcore#24690] [General] [answered]
The issue #13311 is resolved and a PR is merged and included in the 3.0.0 release.

I'm trying to build a module for PiranhaCMS based on **version 3.1.6** of Microsoft.AspNetCore.Authentication.AzureAD.UI to be used optionally instead of the default Piranha.AspNetCore.Identity module. But when I register the services as explained in the [official MS docs](https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-v2-aspnet-core-webapp#startup-class), I run into the same issue as #13311.

**OptionsValidationException: The 'Instance' option must be provided.**

Stacktrace:
```
Microsoft.Extensions.Options.OptionsFactory<TOptions>.Create(string name)
Microsoft.Extensions.Options.OptionsMonitor<TOptions>+<>c__DisplayClass11_0.<Get>b__0()
System.Lazy<T>.ViaFactory(LazyThreadSafetyMode mode)
System.Lazy<T>.ExecutionAndPublication(LazyHelper executionAndPublication, bool useDefaultConstructor)
System.Lazy<T>.CreateValue()
System.Lazy<T>.get_Value()
Microsoft.Extensions.Options.OptionsCache<TOptions>.GetOrAdd(string name, Func<TOptions> createOptions)
Microsoft.Extensions.Options.OptionsMonitor<TOptions>.Get(string name)
Microsoft.AspNetCore.Authentication.AzureAD.UI.AzureADOpenIdConnectOptionsConfiguration.Configure(string name, OpenIdConnectOptions options)
Microsoft.Extensions.Options.OptionsFactory<TOptions>.Create(string name)
Microsoft.Extensions.Options.OptionsMonitor<TOptions>+<>c__DisplayClass11_0.<Get>b__0()
System.Lazy<T>.ViaFactory(LazyThreadSafetyMode mode)
System.Lazy<T>.ExecutionAndPublication(LazyHelper executionAndPublication, bool useDefaultConstructor)
System.Lazy<T>.CreateValue()
System.Lazy<T>.get_Value()
Microsoft.Extensions.Options.OptionsCache<TOptions>.GetOrAdd(string name, Func<TOptions> createOptions)
Microsoft.Extensions.Options.OptionsMonitor<TOptions>.Get(string name)
Microsoft.AspNetCore.Authentication.AuthenticationHandler<TOptions>.InitializeAsync(AuthenticationScheme scheme, HttpContext context)
Microsoft.AspNetCore.Authentication.AuthenticationHandlerProvider.GetHandlerAsync(HttpContext context, string authenticationScheme)
Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
Piranha.AspNetCore.SitemapMiddleware.Invoke(HttpContext context, IApi api, IApplicationService service) in SitemapMiddleware.cs
            await _next.Invoke(context);
Piranha.AspNetCore.StartPageMiddleware.Invoke(HttpContext context, IApi api, IApplicationService service) in StartPageMiddleware.cs
            await _next.Invoke(context);
Piranha.AspNetCore.ArchiveMiddleware.Invoke(HttpContext context, IApi api, IApplicationService service) in ArchiveMiddleware.cs
            await _next.Invoke(context);
Piranha.AspNetCore.PostMiddleware.Invoke(HttpContext context, IApi api, IApplicationService service) in PostMiddleware.cs
            await _next.Invoke(context);
Piranha.AspNetCore.PageMiddleware.Invoke(HttpContext context, IApi api, IApplicationService service) in PageMiddleware.cs
            await _next.Invoke(context);
Piranha.AspNetCore.AliasMiddleware.Invoke(HttpContext context, IApi api, IApplicationService service) in AliasMiddleware.cs
            await _next.Invoke(context);
Piranha.AspNetCore.ApplicationMiddleware.Invoke(HttpContext context, IApi api, IApplicationService service) in ApplicationMiddleware.cs
            await _next.Invoke(context);
Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.Invoke(HttpContext httpContext)
Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider)
Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```

I created this PR as a reference: https://github.com/eloekset/piranha.core/pull/9

And the appsettings.json definitely has the Instance property set:
![image](https://user-images.githubusercontent.com/791721/89690697-66becc00-d907-11ea-94be-67099621e156.png)

I use the **MvcWeb** test application included in the Piranha solution:
![image](https://user-images.githubusercontent.com/791721/89690862-d5038e80-d907-11ea-8d2a-afbdcc55114a.png)

**Comment by tidyui** [comment, marked as answer]
@eloekset Are you sure you’re using the latest version of the docs? Given that the example contains the following line:

    .SetCompatibilityVersion(CompatibilityVersion.Version_2_1);

Is it really compatible with asp.net core 3.1? I know there were a couple of changes affecting startup configuration between 2.x and 3.x, especially regarding the order of how security components should be registered.

**Reply by eloekset** [reply]
To confirm that it works out of the box with a ASP.NET Core 3.1 application, I used the File -> New Project template that configures authentication with Work or School Accounts (although I'm neither at work nor school - I just want to use Azure AD).
![image](https://user-images.githubusercontent.com/791721/89706313-b54e8380-d964-11ea-8337-a34b39cf9d12.png)

I think I'll have to look at how to disable the default Piranha.AspNetCore.Identity module, because there's probably a conflict as mentioned in #13311 or something similar. However it's very difficult to guess what's going on inside these libraries and resolve these kind of issues.

**Reply by tidyui** [reply]
I think in your case you can just remove the a Piranha Identity packages all together. You will loose the management UI, but you’ll manage your users from within Azure AD anyway. Also you’ll need to register a service for the ISecurity interface so Piranha knows how to authenticate users when logging into the manager.

**Reply by eloekset** [reply]
OMG, I overlooked a minor detail with major impact. The options object didn't get bound to the configuration. Commit that fixes the issue: https://github.com/eloekset/piranha.core/pull/9/commits/3d8deeb8017e9b511bd028ac667479dd8fc6f640

And a picture in case I change the branch so the commit link becomes invalid:
![image](https://user-images.githubusercontent.com/791721/89708092-79bbb580-d974-11ea-88d6-f3ebf73e1b54.png)

There's obviously more work to be done to this module, but we'll continue that discussion over at the Piranha project, since it's out of scope of the Azure AD authentication library.
